processor/src/main/java/org/mapstruct/ap/processor/AnnotationBasedComponentModelProcessor.java

```diff
@@ -55,6 +55,7 @@ public Mapper process(ProcessorContext context, TypeElement mapperTypeElement, M
             return mapper;
         }
 
+        mapper.removeDecorator();
```

Comment:
Yes! Well, actually, only in the CDI and Spring implementations. 

I reworked the code a little bit and instead added a method `protected abstract boolean shouldDecoratorBeRemoved();` which all implementations return either `true` or `false` from.